kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/fifo/Fifo.java

```diff
@@ -0,0 +1,237 @@
+
+        public FifoEmitterThread(String threadName, int queueCapacity, boolean discardEnvelopes) {
+            this.queue = new ArrayList<WireEnvelope>();
+            this.queueCapacity = queueCapacity;
+            setName(threadName);
+            if (discardEnvelopes) {
+                submitter = getEnvelopeDiscardingSubmitter();
+            } else {
+                submitter = getEmitterBlockingSubmitter();
+            }
+        }
+
+        private Consumer<WireEnvelope> getEnvelopeDiscardingSubmitter() {
+            return (envelope) -> {
+                try {
+                    lock.lock();
+                    if (!run || queue.size() >= queueCapacity) {
+                        logger.debug("envelope discarded");
+                        return;
+                    } else {
+                        queue.add(envelope);
+                        producer.signal();
+                        logger.debug("envelope submitted");
+                    }
+                } finally {
+                    lock.unlock();
+                }
+            };
+        }
+
+        private Consumer<WireEnvelope> getEmitterBlockingSubmitter() {
+            return (envelope) -> {
+                try {
+                    lock.lock();
+                    while (run && queue.size() >= queueCapacity) {
+                        consumer.await();
+                    }
+                    if (!run) {
+                        return;
+                    }
+                    queue.add(envelope);
+                    producer.signal();
+                    logger.debug("envelope submitted");
+                } catch (InterruptedException e) {
+                    logger.warn(message.fifoInterruptedWhileSubmitting(), e);
+                } finally {
+                    lock.unlock();
+                }
+            };
+        }
+
+        public void shutdown() {
+            try {
+                lock.lock();
+                run = false;
+                producer.signalAll();
+                consumer.signalAll();
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        public void sumit(WireEnvelope envelope) {
+            submitter.accept(envelope);
+        }
+
+        @Override
+        public void run() {
+            while (run) {
+                try {
+                    WireEnvelope next = null;
+                    try {
+                        lock.lock();
+                        while (run && queue.isEmpty()) {
+                            producer.await();
+                        }
+                        if (!run) {
+                            break;
+                        }
+                        next = queue.remove(queue.size() - 1);
+                        consumer.signal();
+                    } finally {
+                        lock.unlock();
+                    }
+                    wireSupport.emit(next.getRecords());
```

Comment:
I believe you can also put `WireSupport.emit(..)` inside previous critical section and add finally to the outer `try..catch`